### PR TITLE
Add typespecs for exception module

### PR DESCRIPTION
**Pull request by solar05**

### CODE REVIEW

File: lib/elixir/lib/exception.ex [hunk 1]
```diff
@@ -767,6 +772,7 @@ defmodule Exception do
       ""
 
   """
+  @spec format_file_line_column(term, term, term, String.t()) :: String.t()
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:775]:
```suggestion
  @spec format_file_line_column(String.t | nil, non_neg_integer | nil, non_neg_integer | nil, String.t()) :: String.t()
```

Similar changes necessary above. :)

File: lib/elixir/lib/exception.ex [hunk 2]
```diff
@@ -673,6 +675,7 @@ defmodule Exception do
       #=> "#Function<...>/1"
 
   """
+  @spec format_fa(fun, non_neg_integer) :: String.t()
```
Comment (whatyouhide) [inline lib/elixir/lib/exception.ex:678]: Should we use this here?

```suggestion
  @spec format_fa(fun, arity) :: String.t()
```

File: lib/elixir/lib/exception.ex [hunk 3]
```diff
@@ -697,6 +700,7 @@ defmodule Exception do
   where func is the name of the enclosing function. Convert to
   "anonymous fn in func/arity"
   """
+  @spec format_mfa(module, fun, non_neg_integer) :: String.t()
```
Comment (whatyouhide) [inline lib/elixir/lib/exception.ex:703]: ...and here?

File: lib/elixir/lib/exception.ex [hunk 4]
```diff
@@ -740,7 +740,7 @@ defmodule Exception do
       ""
 
   """
-  @spec format_file_line(term, term, String.t()) :: String.t()
+  @spec format_file_line(String.t() | nil, non_neg_integer | nil, String.t()) :: String.t()
```
Comment (whatyouhide) [inline lib/elixir/lib/exception.ex:743]: I don't think `line` or `column` can be `0`, so I think we want

```suggestion
  @spec format_file_line(String.t() | nil, pos_integer | nil, String.t()) :: String.t()
```

here and below.

Comment (josevalim) [reply lib/elixir/lib/exception.ex]: they can be zero, line 0 usually means we don't have a precise line too (it is used by erlang in places where a line must be given).

File: lib/elixir/lib/exception.ex [hunk 5]
```diff
@@ -736,6 +740,7 @@ defmodule Exception do
       ""
 
   """
+  @spec format_file_line(String.t() | nil, pos_integer | nil, String.t()) :: String.t()
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:743]:
```suggestion
  @spec format_file_line(String.t() | nil, non_neg_integer | nil, String.t()) :: String.t()
```

File: lib/elixir/lib/exception.ex [hunk 6]
```diff
@@ -767,6 +772,12 @@ defmodule Exception do
       ""
 
   """
+  @spec format_file_line_column(
+          String.t() | nil,
+          pos_integer | nil,
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex:777]:
```suggestion
          non_neg_integer | nil,
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**